File: android/src/main/java/com/github/wuxudong/rncharts/charts/DynamicChartDateFormatter.java
package com.github.wuxudong.rncharts.charts;

import com.github.mikephil.charting.components.AxisBase;
import com.github.mikephil.charting.formatter.ValueFormatter;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;

public class DynamicChartDateFormatter extends ValueFormatter {
    private SimpleDateFormat mSimpleDateFormat;
    private Calendar mCalendar = Calendar.getInstance();

    long[] dates;

    public DynamicChartDateFormatter(long[] dates, Locale locale) {
        this.dates = dates;
        this.mSimpleDateFormat = new SimpleDateFormat("HH:mm", locale);
    }

    @Override
    public String getAxisLabel(float value, AxisBase axis) {
        int dateIndex = Math.abs(Math.round(value));

        if (dateIndex < 0 || dateIndex >= dates.length)
            return "";

        Date date = new Date(dates[dateIndex]);

        int entryIndex = getEntryIndexForValue(value, axis);

        if (entryIndex < 0) {
            return "";
        }

        if (entryIndex == 0) {
            int entryInterval = (int) Math.abs(axis.mEntries[1] - axis.mEntries[0]);
            int previousDateIndex = dateIndex + entryInterval;

            return formatByPreviousDateIndex(previousDateIndex, date);
        }

        int previousDateIndex = Math.abs(Math.round(axis.mEntries[entryIndex - 1]));
        return formatByPreviousDateIndex(previousDateIndex, date);
    }

    private String formatByPreviousDateIndex(int previousDateIndex, Date date) {
        if (previousDateIndex < 0 || previousDateIndex >= dates.length)
            return "";

        Date previousDate = new Date(dates[previousDateIndex]);

        getFormatPattern(date, previousDate);

        return mSimpleDateFormat.format(date);
    }

    private int getEntryIndexForValue(float value, AxisBase axis) {
        for (int i = 0; i < axis.mEntryCount; i++) {
            if (value == axis.mEntries[i]) {
                return i;
            }
        }
        return -1;
    }

    // If value diff > year
    // return year number
    // If value diff > month
    // return month number
    // If value diff > day
    // return day number
    // else
    // return HH:mm
    private void getFormatPattern(Date date1, Date date2) {
        mCalendar.setTime(date1);
        int year1 = mCalendar.get(Calendar.YEAR);
        int month1 = mCalendar.get(Calendar.MONTH);
        int day1 = mCalendar.get(Calendar.DAY_OF_MONTH);
        mCalendar.setTime(date2);
        int year2 = mCalendar.get(Calendar.YEAR);
        int month2 = mCalendar.get(Calendar.MONTH);
        int day2 = mCalendar.get(Calendar.DAY_OF_MONTH);

        if (year1 != year2) {
            mSimpleDateFormat.applyPattern("yyyy");
        } else if (month1 != month2) {
            mSimpleDateFormat.applyPattern("MMM");
        } else if (day1 != day2) {
            mSimpleDateFormat.applyPattern("d");
        } else {
            mSimpleDateFormat.applyPattern("HH:mm");
        }
    }
}

File: ios/ReactNativeCharts/DynamicChartDateFormatter.swift
//
//  DynamicChartDateFormatter.swift
//  ReactNativeCharts
//
//  Created by Taylor Johnson on 6/5/20.
//

import Foundation
import Charts

extension Date {
    func isEqual(to date: Date, toGranularity component: Calendar.Component, in calendar: Calendar = .current) -> Bool {
        calendar.isDate(self, equalTo: date, toGranularity: component)
    }

    func isInSameYear(as date: Date) -> Bool { isEqual(to: date, toGranularity: .year) }
    func isInSameMonth(as date: Date) -> Bool { isEqual(to: date, toGranularity: .month) }
    func isInSameWeek(as date: Date) -> Bool { isEqual(to: date, toGranularity: .weekOfYear) }

    func isInSameDay(as date: Date) -> Bool { Calendar.current.isDate(self, inSameDayAs: date) }
}

// assumes values are -Index
// TODO take timeUnit
// TODO clean up
open class DynamicChartDateFormatter: NSObject, IValueFormatter, IAxisValueFormatter {
    
    open var dateFormatter = DateFormatter()
    
    open var dates = [Double]()
    
    public override init() {
        
    }
    
    public init(dates: [Double], locale: String?) {
        self.dates = dates
        self.dateFormatter.locale = Locale(identifier: locale ?? Locale.current.languageCode ?? "en_US")
    }

    open func stringForValue(_ value: Double, axis: AxisBase?) -> String {
        let dateIndex = -Int(value.rounded())
        guard dates.indices.contains(dateIndex), dateIndex == -Int(value) else { return "" }
        
        let date = Date(timeIntervalSince1970: dates[dateIndex] / 1000)
        
        let entries = axis?.entries ?? []
        guard let entryIndex = entries.firstIndex(of: value) else { return "" }

        if (entryIndex == 0) {
            let entryInterval = abs(Int(entries[1] - entries[0]))
            let previousDateIndex = dateIndex + entryInterval
            return formatByPreviousDateIndex(previousDateIndex, date: date)
        }
        
        let previousDateIndex = -Int(entries[entryIndex - 1].rounded())
        return formatByPreviousDateIndex(previousDateIndex, date: date)
    }
    
    public func stringForValue(_ value: Double, entry: ChartDataEntry, dataSetIndex: Int, viewPortHandler: ViewPortHandler?) -> String {
        // TODO impl
        return ""
    }

    private func formatByPreviousDateIndex(_ previousDateIndex: Int, date: Date) -> String {
        guard dates.indices.contains(previousDateIndex) else { return "" }
        let previousDate = Date(timeIntervalSince1970: dates[previousDateIndex] / 1000)

        updateFormatting(date1: date, date2: previousDate)

        return dateFormatter.string(from: date)
    }
    
    // If value diff > year
        // return year number
    // If value diff > month
        // return month number
    // If value diff > day
        // return day number
    // else
        // return HH:mm
    private func updateFormatting(date1: Date, date2: Date) {
        if (!date1.isInSameYear(as: date2)) {
            dateFormatter.dateFormat = "yyyy"
        } else if (!date1.isInSameMonth(as: date2)) {
            dateFormatter.dateFormat = "MMM"
        } else if (!date1.isInSameDay(as: date2)) {
            dateFormatter.dateFormat = "d"
        } else {
            dateFormatter.dateFormat = "HH:mm"
        }
    }
    
}
